osutil/sync_dir.go

```diff
@@ -0,0 +1,158 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"syscall"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+	UID     uint32
+	GID     uint32
+}
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions and ownership are what is desired. Unexpected
+// files are removed.  Missing files are created and differing files are
+// corrected.  Files not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
+// returned to the caller for any extra processing that might be required (e.g.
+// to run some helper program).
+//
+// The return value is: created, corrected, removed []string, err error
+func EnsureDirState(dir, glob string, content map[string]*FileState) ([]string, []string, []string, error) {
+	matches, err := filepath.Glob(path.Join(dir, glob))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	found := make(map[string]bool)
+	var created, corrected, removed []string
+	// Analyze files that inhabit the subset defined by our glob pattern.
+	for _, name := range matches {
+		baseName := path.Base(name)
+		// Remove files that should not be here.
+		var expected *FileState
+		var shouldBeHere bool
+		if expected, shouldBeHere = content[baseName]; !shouldBeHere {
```

Comment:
I think this goes back to golint giving us exactly the opposite advice. Personally I find it tedious that we have to make golint happy this way when it flies in the face of what is more natural. I don't quite know how to convince golint to do better or start the process to evict it from having a say in our mergability.